<?xml version="1.0"?>
<launch>
  <!-- arguments -->
  <arg name="nodelet_manager" default="velodyne_nodelet_manager" />
  <arg name="enable_floor_detection" default="true" />
  
  <node pkg="nodelet" type="nodelet" name="$(arg nodelet_manager)" args="manager" output="screen"/>
  
  <node pkg="tf" type="static_transform_publisher" name="lidar_frame_publisher" args="0 0 0.81 1.5707963 0 0 base_link velodyne 100" />
  <node pkg="tf" type="static_transform_publisher" name="gps_frame_publisher"   args="0 -0.5 0 0 0 0   base_link gps 100" />
  
  <!-- prefiltering_nodelet -->
  <node pkg="nodelet" type="nodelet" name="prefiltering_nodelet" args="load hdl_graph_slam/PrefilteringNodelet $(arg nodelet_manager)">
	<param name="raw_points_topic" value="/velodyne_points" />
    <!-- in case base_link_frame is blank, mapping will be performed in the lidar frame -->
    <param name="base_link_frame" value="base_link" />
    <!-- distance filter -->
    <param name="use_distance_filter" value="true" />
    <param name="distance_near_thresh" value="2.0" />
    <param name="distance_far_thresh" value="30.0" />
    <!-- NONE, VOXELGRID, or APPROX_VOXELGRID -->
    <param name="downsample_method" value="VOXELGRID" />
    <param name="downsample_resolution" value="0.1" />
    <!-- NONE, RADIUS, or STATISTICAL -->
    <param name="outlier_removal_method" value="RADIUS" />
    <param name="statistical_mean_k" value="30" />
    <param name="statistical_stddev" value="1.2" />
    <param name="radius_radius" value="0.5" />
    <param name="radius_min_neighbors" value="10" />
  </node>

 <!-- floor_detection_nodelet -->
  <node pkg="nodelet" type="nodelet" name="floor_detection_nodelet" args="load hdl_graph_slam/FloorDetectionNodelet $(arg nodelet_manager)" if="$(arg enable_floor_detection)">
    <param name="points_topic" value="/prefiltering/filtered_points"/>
    <param name="tilt_deg" value="0.0" />
    <param name="sensor_height" value="0.81" />
    <param name="height_clip_range" value="0.2" />
    <param name="floor_pts_thresh" value="256" />
    <param name="use_normal_filtering" value="true" />
    <param name="normal_filter_thresh" value="20.0" />
  </node>

  <node pkg="nodelet" type="nodelet" name="location_nodelet" args="load hdl_graph_slam/LocationNodelet $(arg nodelet_manager)">
    <param name="utm_topic" value="/gps_odom" />
    <param name="points_topic" value="/floor_detection/floor_filtered_points" />
    <param name="odom_frame_id" value="odom" />
    <param name="downsample_method" value="NONE" />
    <param name="downsample_resolution" value="0.1" />
    <!-- ICP, GICP, NDT, GICP_OMP, or NDT_OMP(recommended) -->
    <param name="registration_method" value="NDT_OMP" />
    <param name="ndt_resolution" value="1.0" />
    <param name="ndt_num_threads" value="0" />
    <param name="ndt_nn_search_method" value="DIRECT7" />
    <param name="keyframes_directory" value="/home/zwei/wendao/map/1_2019-09-27-14-04-43" />
    
  </node>
  
  <node pkg="hdl_graph_slam" type="record_odom_node" name="record_odom_node">
  	 <param name="file_name" value="/home/zwei/wendao/map/1_2019-09-27-14-04-43/odom.txt"/>
  	 <param name="gps_odom_topic" value="/gps_odom"/>
  	 <param name="slam_odom_topic" value="/location/odom"/>

  </node>


</launch>
